src/python/pants/rules/core/run.py

```diff
@@ -1,35 +1,56 @@
 # Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).
 
+from pathlib import Path
+
+from pants.build_graph.address import Address
 from pants.engine.addressable import BuildFileAddresses
 from pants.engine.console import Console
+from pants.engine.fs import DirectoryToMaterialize, Workspace
 from pants.engine.goal import Goal
 from pants.engine.interactive_runner import InteractiveProcessRequest, InteractiveRunner
 from pants.engine.rules import console_rule
+from pants.engine.selectors import Get
+from pants.rules.core.binary import CreatedBinary
+from pants.util.contextutil import temporary_dir
 
 
 class Run(Goal):
   """Runs a runnable target."""
-  name = 'v2-run'
+  name = 'run'
 
 
 @console_rule
-def run(console: Console, runner: InteractiveRunner, build_file_addresses: BuildFileAddresses) -> Run:
-  console.write_stdout("Running the `run` goal\n")
-
-  request = InteractiveProcessRequest(
-    argv=["/usr/bin/python"],
-    env=("TEST_ENV", "TEST"),
-    run_in_workspace=False,
-  )
-
-  try:
-    res = runner.run_local_interactive_process(request)
-    print(f"Subprocess exited with result: {res.process_exit_code}")
-    yield Run(res.process_exit_code)
-  except Exception as e:
-    print(f"Exception when running local interactive process: {e}")
-    yield Run(-1)
+def run(console: Console, workspace: Workspace, runner: InteractiveRunner, addresses: BuildFileAddresses) -> Run:
+  binaries = yield [Get(CreatedBinary, Address, address.to_address()) for address in addresses]
+
+  exit_codes = []
+  for (binary, bfa) in zip(binaries, addresses):
+    target = bfa.to_address()
+    with temporary_dir(cleanup=False) as tmpdir:
+      dirs_to_materialize = (DirectoryToMaterialize(path=str(tmpdir), directory_digest=binary.digest),)
+      workspace.materialize_directories(dirs_to_materialize)
+
+      console.write_stdout(f"Running target: {target}\n")
+      full_path = str(Path(tmpdir, binary.command))
+      run_request = InteractiveProcessRequest(
+        argv=[full_path],
```

Comment:
Will want to add a TODO about adding passthru args. I just created https://github.com/pantsbuild/pants/issues/8579 for this.